File: src/com/themaskedcrusader/tmcz/modules/thirst/ThirstSchedule.java
/*
 * Copyright 2013 Topher Donovan (themaskedcrusader.com)
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package com.themaskedcrusader.tmcz.modules.thirst;

import com.themaskedcrusader.bukkit.config.Messages;
import com.themaskedcrusader.bukkit.config.Settings;
import org.bukkit.ChatColor;
import org.bukkit.entity.Player;
import org.bukkit.plugin.java.JavaPlugin;

public class ThirstSchedule {
    private static JavaPlugin plugin;

    public ThirstSchedule(JavaPlugin plugin) {
        ThirstSchedule.plugin = plugin;

        // Schedule to calculate thirst
        plugin.getServer().getScheduler().scheduleSyncRepeatingTask(plugin, new Runnable() {
            public void run() {
                ThirstSchedule.calculateThirst();
            }
        }, 20L, Settings.getConfig().getLong(Thirst.TICKS));

        // Schedule to damage due to thirst
        plugin.getServer().getScheduler().scheduleSyncRepeatingTask(plugin, new Runnable() {
            public void run() {
                ThirstSchedule.calculateThirstDamage();
            }
        }, 20L, Settings.getConfig().getLong(Thirst.DAMAGE_TICKS));
    }


    public static void calculateThirst() {
        Player[] players = plugin.getServer().getOnlinePlayers();
        for (Player player : players) {
            if (Thirst.isAllowed(player) && player.getLevel() > 0) {
                player.setLevel(player.getLevel() - 1);
                sendPlayerThirstMessage(player);
            }
        }
    }

    public static void calculateThirstDamage() {
        Player[] players = plugin.getServer().getOnlinePlayers();
        for (Player player : players) {
            if (Thirst.isAllowed(player) && player.getLevel() == 0) {
                if (!player.isOp() || !Settings.getConfig().getBoolean("world.op-is-god")) {
                    player.damage(Settings.getConfig().getInt(Thirst.DAMAGE_HIT));
                }
                if (player.getHealth() == Settings.getConfig().getInt(Thirst.DEATH_LVL)) {
                    player.sendMessage(ChatColor.RED + Messages.getConfig().getString(Thirst.DEATH_MSG));
                }
            }
        }
    }

    private static void sendPlayerThirstMessage(Player player) {
        String message = "" + ChatColor.YELLOW;
        if (player.getLevel() == Settings.getConfig().getInt(Thirst.PARCH_1)) {
            player.sendMessage(message + Messages.getConfig().getString(Thirst.PARCH_1_MSG));
        } else if (player.getLevel() == Settings.getConfig().getInt(Thirst.PARCH_2)) {
            player.sendMessage(message + Messages.getConfig().getString(Thirst.PARCH_2_MSG));
        } else if (player.getLevel() == Settings.getConfig().getInt(Thirst.PARCH_3)) {
            player.sendMessage(message + Messages.getConfig().getString(Thirst.PARCH_3_MSG));
        } else if (player.getLevel() == Settings.getConfig().getInt(Thirst.PARCH_4)) {
            player.sendMessage(message + Messages.getConfig().getString(Thirst.PARCH_4_MSG));
        }
    }
}

File: src/main/kotlin/me/schalk/spigot/tmcz/modules/thirst/ThirstSchedule.kt
/*
 * Copyright (c) 2013 - 2023 Christopher Schalk
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package me.schalk.spigot.tmcz.modules.thirst

import me.schalk.spigot.lib.config.getSettings
import me.schalk.spigot.lib.serializer.abovePlayerHead
import me.schalk.spigot.tmcz.data.GameData
import org.bukkit.ChatColor
import org.bukkit.Particle
import org.bukkit.entity.Player
import org.bukkit.plugin.java.JavaPlugin
import org.bukkit.potion.PotionEffect
import org.bukkit.potion.PotionEffectType

class ThirstSchedule(plugin: JavaPlugin) : ThirstModule() {
    
    init{
        ThirstSchedule.plugin = plugin
        plugin.server.scheduler.scheduleSyncRepeatingTask(plugin, { calculateThirst() }, 20L, TICKS )
        plugin.server.scheduler.scheduleSyncRepeatingTask(plugin, { calculateThirstDamage() }, 20L, DAMAGE_TICKS)
    }

    companion object {
        private lateinit var plugin: JavaPlugin

        fun calculateThirst() {
            plugin.server.onlinePlayers.forEach{ player ->
                if (isAllowed(player) && player.level > 0) {
                    player.level = player.level - 1
                    sendPlayerThirstMessage(player)
                    plugin.server.scheduler.runTaskLater(plugin, Runnable {
                        val location = abovePlayerHead(player.location)
                        player.spawnParticle(Particle.BUBBLE_POP, location.x, location.y, location.z, 1, 0.2, 0.2, 0.2, 0.01)
                    }, 5)
                }
            }
        }

        fun calculateThirstDamage() {
            plugin.server.onlinePlayers.forEach{ player ->
                if (isAllowed(player) && player.level == 0) {
                    GameData.getPlayer(player).thirsty = true
                    if (!player.isOp || !getSettings().getConfig().getBoolean("world.op-is-god")) { // TODO move to world config
                        player.damage(DAMAGE_HIT)
                        val location = abovePlayerHead(player.location)
                        player.spawnParticle(Particle.BUBBLE_POP, location.x, location.y, location.z, 3, 0.3, 0.3, 0.3, 0.01)
                    }
                    if (player.health <= 4.0) {
                        if (!player.hasPotionEffect(PotionEffectType.CONFUSION)) {
                            player.addPotionEffect(PotionEffect(PotionEffectType.CONFUSION,500, 255, false, false, false))
                        }
                    }
                }
            }
        }

        private fun sendPlayerThirstMessage(player: Player) {
            val yellow = ChatColor.YELLOW.toString()
            when (player.level) {
                PARCH_1 -> { player.sendMessage(yellow + PARCH_1_MSG) }
                PARCH_2 -> { player.sendMessage(yellow + PARCH_2_MSG) }
                PARCH_3 -> { player.sendMessage(yellow + PARCH_3_MSG) }
                PARCH_4 -> { player.sendMessage(yellow + PARCH_4_MSG) }
            }
        }
    }
}